Combat toolgun

Kilometres opened the pull request:
~Adds my combat toolgun SWEP.~
Adds a new SWEP that I made, which looks like a toolgun but when fired shoots using the toolgun tracer and sounds. It has a primary fire similar to an automatic pistol and a secondary fire similar to a shotgun spread. I've balanced the weapon to ensure it does not affect any PvP metas. This SWEP is a bit of a meme weapon as it looks and sounds like the standard toolgun but hurts a lot more.

Kilometres committed:
Create kilos_combat_toolgun.lua

Added Weapon Lua File

Kilometres committed:
Update kilos_combat_toolgun.lua

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+
+SWEP.Base = "weapon_base"
+
+local ShootSound = Sound("weapons/airboat/airboat_gun_lastshot2.wav")
```
I think this'd be better as a table entry into the SWEP, and the sound path turned into a constant.

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+SWEP.ViewModelFOV		= 60
+SWEP.ViewModel			= "models/weapons/c_toolgun.mdl"
+SWEP.WorldModel			= "models/weapons/w_toolgun.mdl"
+SWEP.UseHands           = true
```
Inconsistent formatting with the rest of the file and TAS' template stylua file

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+
+SWEP.FiresUnderwater = true
+SWEP.DrawWeaponInfoBox = false
+SWEP.ReloadSound = "weapons/airboat/airboat_gun_lastshot2.wav"
```
Could reference the constant I suggested for this path

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+function SWEP:Initialize()
+    util.PrecacheSound(ShootSound) 
+    util.PrecacheSound(self.ReloadSound) 
+    self:SetWeaponHoldType( self.HoldType )
```
Why are the call parentheses spaced?

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+
+function SWEP:PrimaryAttack()
+
+    if ( !self:CanPrimaryAttack() ) then return end
```
Spaced parentheses and a C operator...

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+    bullet.Num = self.Primary.NumberofShots
+    bullet.Src = self.Owner:GetShootPos()
+    bullet.Dir = self.Owner:GetAimVector()
+    bullet.Spread = Vector( self.Primary.Spread * 0.1 , self.Primary.Spread * 0.1, 0)
```
More spaced parentheses, and the spaced comma also doesn't look right

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+    bullet.Distance = 100000
+        
+    local rnda = self.Primary.Recoil * -1
+    local rndb = self.Primary.Recoil * math.random(-1, 1)
```
I think these variables could be a little more descriptive

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+    bullet.Force = self.Primary.Force
+    bullet.Damage = self.Primary.Damage
+    bullet.AmmoType = self.Primary.Ammo
+    bullet.Distance = 100000
```
Could use scientific notation here, `1e5`

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+        
+    self:ShootEffects()
+        
+    self.Owner:FireBullets( bullet )
```
Spaced parentheses

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+        
+    self.Owner:FireBullets( bullet )
+    self:EmitSound(ShootSound)
+    self.Owner:ViewPunch( Angle( rnda,rndb,rnda ) )
```
More spaced parentheses

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+    self:EmitSound(ShootSound)
+    self.Owner:ViewPunch( Angle( rnda,rndb,rnda ) )
+
+    self:SetNextPrimaryFire( CurTime() + self.Primary.Delay )
```
Another spaced parentheses

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+    bullet.Num = self.Secondary.NumberofShots
+    bullet.Src = self.Owner:GetShootPos()
+    bullet.Dir = self.Owner:GetAimVector()
+    bullet.Spread = Vector( self.Secondary.Spread * 0.1 , self.Secondary.Spread * 0.1, 0)
```
nit: You could extract this bullet logic into another function that makes a bullet

Otherwise everything in the primary attack function is also an issue here

yogwoggf commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,127 @@
+
+if !SERVER then
+    killicon.Add( "kilos_combat_toolgun", "vgui/face/grin", Color( 255, 255, 255, 255 ) )
+    SWEP.WepSelectIcon = surface.GetTextureID( 'vgui/gmod_tool' )
```
Spaced parentheses and sudden usage of single quotes?

yogwoggf reviewed (commented):
I'm not that involved in TAS GMod right now, nor am I knowledgeable of this new SWEP repo, but here are some changes that I would do. Not going to deny the current PR or accept it because again I dont know how the new swep repo works. Pretty much just my personal feedback

Derpius commented:
@yogwoggf there’s no reason to bloat the pr with formatting comments that have been covered by CI anyway

@Kilometres you should install the stylua and gluafixer vscode extensions locally per the z2h in the readme. Formatting on save should be automatically enabled by `.vscode/settings.json`

Derpius commented:
Also @yogwoggf I recommend reading https://google.github.io/eng-practices/review/reviewer/comments.html (most of your comments are nitpicks and should really be marked as such)

Derpius commented:
Had an actual quick look through now (not a full review though given the failing lint) and generally seems fine

Could you add a more detailed description to the PR please? The comments in the template should give some guidance on that

Kilometres commented:
Yep, I'll keep in mind all of the comments and I'll update it later today.

Kilometres committed:
Update kilos_combat_toolgun.lua

- Fixed inconsistencies in formatting and code
- Broken Primary/Secondary Fire into two functions, one to create the bullet object, one to shoot primary/secondary.

Kilometres committed:
Minor: Update kilos_combat_toolgun.lua

Kilometres committed:
Update kilos_combat_toolgun.lua lint

Derpius committed:
Merge branch 'master' into combat_toolgun

Derpius committed:
Apply stylua config to combat toolgun

Kilometres commented:
Updated description to provide slightly more detail on the PR.

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	bullet.AmmoType = fire.Ammo
+	bullet.Distance = 1e5
+	bullet.RecoilPR = fire.Recoil * -1 --Recoil for Pitch & Roll
+	bullet.recoilY = fire.Recoil * math.random(-1, 1) --Recoil for Yaw
```
Do you mean `math.Rand` here?

`math.random` will return an integer when called with arguments (returns uniform `[0,1)` float otherwise), but that might have been intentional?

Kilometres replied:
Yep intentional, wanted to randomise the direction, but constant magnitude

Derpius replied:
In that case I believe the magnitude will be `0` a third of the time

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	self:SetWeaponHoldType(self.HoldType)
+end
+
+function SWEP:createBullet(fire)
```
nit: I think we could use a more descriptive parameter name here and elsewhere. Maybe `bulletProperties` or `bulletCreateOptions`?

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	self:ShootEffects()
+	self:GetOwner():FireBullets(bullet)
+	self:EmitSound(self.ShootSound)
+	self:GetOwner():ViewPunch(Angle(bullet.RecoilPR, bullet.recoilY, bullet.RecoilPR))
```
Looks like these recoil props aren't part of the `Bullet` structure. Could we avoid using abbreviations and keep the casing consistent please?

Also, I don't think we actually need to store these properties on the bullet object given they're not used elsewhere and we have access to `fire` here?

Derpius replied:
Unrelated to the above, but I'm also not sure rolling the camera when applying recoil is either a good user experience or physically accurate

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	bullet.Num = fire.NumberofShots
+	bullet.Src = self:GetOwner():GetShootPos()
+	bullet.Dir = self:GetOwner():GetAimVector()
+	bullet.Spread = Vector(fire.Spread * 0.1, fire.Spread * 0.1, 0)
```
Why are we multiplying by `0.1` here?

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	bullet.Force = fire.Force
+	bullet.Damage = fire.Damage
+	bullet.AmmoType = fire.Ammo
+	bullet.Distance = 1e5
```
The default distance [on the wiki](https://wiki.facepunch.com/gmod/Structures/Bullet) seems fine to me, do we need to overwrite it?

Kilometres replied:
I changed it to match the toolgun.

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+	self:SetNextSecondaryFire(CurTime() + self.Secondary.Delay)
+end
+
+if not SERVER then
```
nit: `if CLIENT then`

Kilometres replied:
Fixed

Derpius commented on lua/weapons/kilos_combat_toolgun.lua:
```diff
@@ -0,0 +1,105 @@
+end
+
+if not SERVER then
+	killicon.Add("kilos_combat_toolgun", "vgui/face/grin", Color(255, 255, 255, 255))
```
I imagine this kill icon isn't the most appropriate for the weapon, is there an icon of the toolgun at all? (that's suitable to be used as a kill icon)

Otherwise we can leave this as a placeholder and add a toolgun kill icon in the future

Kilometres replied:
I've left it for now, but i do have images for both the HUD weapon select and kill icon that fit the theme of vanilla HL2 I just need to add them.

Derpius reviewed (commented):
Looks good for the most part, mostly minor comments

Kilometres committed:
Update kilos_combat_toolgun.lua

Added Derpius' comments.

Kilometres commented:
Added your suggestions. If there are any comments you made that I resolved remake them and I won't do that again :)

Derpius reviewed (approved):
LGTM - Left that one comment open on the variable naming that I mentioned in Discord, but this looks good to QA now 

I'll do the QA given it's fairly simple and should give a decent example to others doing QA in the future

Looks like there's a small formatter issue. I think Stylua isn't configured correctly in your editor? (an issue I've had before is LuaLS's formatter overriding it)

Kilometres committed:
Minor: Updated parameter names

Any comments about my illiteracy will be treated as discrimination and I will take action under the Equality Act 2010.

Kilometres committed:
Minor: Formatting

Derpius commented:
# QA Result: Failed ❌ 

Looks like the `AddCSLuaFile` call wasn't required, as it seems to network fine when testing with a local server

## ❌ Blockers

- The weapon has a valid hold type for the world model ❌ 
  - The hold type for the world model is unset, leaving GMod to default to the camera hold type
![image](https://github.com/TAServers/sweps/assets/49565664/960e5722-9eda-4fcc-a433-519f2a54fc55)


## ⚠️ Warnings

- Primary fire
  - Applies recoil ⚠️
    - Yes, but a very small amount
    - Should be fine to merge, but one to watch for user feedback
    - I'd try this one yourself first though as it's even less than the 9mm
- Secondary fire
  - Applies recoil ⚠️
    - Pitches the camera down which is a little unexpected
    - Also, the binary choice between left and right yaw is very noticeable
    - Again not a blocker, but one to watch for user feedback
- The time-to-kill for the primary and secondary fire modes are different ⚠️ 
  - It takes 5 shots with either mode to kill an NPC, which isn't expected (for me)
  - This is assuming all shots with the shotgun blast hit
  - Might be worth increasing the secondary fire damage? 

## ✅ Passed

- Primary fire
  - Fires a single projectile ✅ 
  - Plays toolgun sounds ✅ 
  - Fires a toolgun tracer ✅ 
  - Fires automatically, with a delay ✅
- Secondary fire
  - Fires multiple projectiles with a shotgun spread ✅ 
  - Plays toolgun sounds ✅ 
  - Fires a toolgun tracer for each projectile ✅ 
  - Does not fire automatically ✅ 
- The weapon is placed in the expected slot in the weapon selector (pistols) ✅ 
- The weapon has suitable spawn menu, weapon select and kill icon textures ✅ 
  - Kill icon texture is a known placeholder
- The view model is correct ✅ 
- The world model is correct ✅ 
- The view model hold type is correct ✅

Kilometres committed:
Minor: HoldType Capitalisation

Kilometres commented:
# QA Resolution

## ❌ Blockers

- The weapon has a valid hold type for the world model ❌ 
  - **R:** My illiteracy strikes again, the hold type was `Pistol` where it should've been `pistol`.


## ⚠️ Warnings

- The time-to-kill for the primary and secondary fire modes are different ⚠️ 
  - It takes 5 shots with either mode to kill an NPC, which isn't expected (for me)
  - This is assuming all shots with the shotgun blast hit
  - Might be worth increasing the secondary fire damage? 
    -  **R:** I attempted to balance it as best as possible and was unsure of the secondary fire (as I didn't want it to be an important PvP weapon just a bit of a meme) so I thought it would be safer to undershoot the damage. I expect feedback and will adjust when an adequate amount is received (same with the other ⚠️ warnings)

Derpius commented:
# Re-QA Result: Passed ✅

- The weapon has a valid hold type for the world model ✅
